source/common/upstream/upstream_impl.cc

```diff
@@ -1160,12 +1161,18 @@ void ClusterImplBase::onInitDone() {
       pending_initialize_health_checks_ += host_set->hosts().size();
     }
 
+    // We use a weak pointer to callback_lifetime_guard_ to ensure that we don't segfault if the
+    // callback runs after 'this' is deleted.
+    std::weak_ptr<uint32_t> lifetime_guard = callback_lifetime_guard_;
     // TODO(mattklein123): Remove this callback when done.
-    health_checker_->addHostCheckCompleteCb([this](HostSharedPtr, HealthTransition) -> void {
-      if (pending_initialize_health_checks_ > 0 && --pending_initialize_health_checks_ == 0) {
-        finishInitialization();
-      }
-    });
+    health_checker_->addHostCheckCompleteCb(
```

Comment:
Thanks for the feedback! It turns out it was simple to clear the callbacks list from the health checker early in its deletion to prevent the callback from running in the first place.